Improve autoloading_and_reloading_constants.md [ci-skip] (#44598)
rails/rails: guides/source/autoloading_and_reloading_constants.md

WARNING. Please do not mutate `ActiveSupport::Dependencies.autoload_paths`; the public interface to change autoload paths is `config.autoload_paths`.

WARNING: You cannot autoload code in the autoload paths while the application boots. In particular, directly in `config/initializers/*.rb`. Please check [_Autoloading when the application boots_](#autoloading-when-the-application-boots) down below for valid ways to do that.

The autoload paths are managed by the `Rails.autoloaders.main` autoloader.

config.autoload_once_paths
--------------------------

> It may be wanted to autoload classes and modules without reloading them. The autoload once paths store code that can be autoloaded, but won't be reloaded.

Reviewer: ```suggestion
It may be desired to autoload classes and modules without reloading them. The autoload once paths store code that can be autoloaded, but won't be reloaded.
```

> It may be desired to autoload classes and modules without reloading them. The autoload once paths store code that can be autoloaded, but won't be reloaded.

Author: Thank you @p8, here's the clearer suggestion to your suggestion:

```suggestion
It may be desired to autoload classes and modules without reloading them. The `autoload_once_paths` configuration stores code that can be autoloaded, but won't be reloaded.
```

What do you think?

Making `MoneySerializer` reloadable would be confusing, because reloading an edited version would have no effect on that class object stored in Active Job. Indeed, if `MoneySerializer` was reloadable, starting with Rails 7 such initializer would raise a `NameError`.

> Another use case is engines decorating framework classes:

Reviewer: ```suggestion
Another use case is when engines decorate framework classes:
```

$LOAD_PATH{#load_path}
----------

> Autoload paths are added to `$LOAD_PATH` by default. However, Zeitwerk uses absolute file names internally, and the application should not issue `require` calls for autoloadable files, so those directories are actually not needed there. There is an opt-out flag:

Reviewer: ```suggestion
Autoload paths are added to `$LOAD_PATH` by default. However, Zeitwerk uses absolute file names internally, and the application should not issue `require` calls for autoloadable files, so those directories are actually not needed there. This behavior can be disabled with a flag:
```

```ruby
config.add_autoload_paths_to_load_path = false
```

> That may speed up legitimate `require` calls a bit since there are fewer lookups. Also, if the application uses [Bootsnap](https://github.com/Shopify/bootsnap), that saves the library from building unnecessary indexes, and saves the memory space they need.

Reviewer: ```suggestion
That may speed up legitimate `require` calls a bit since there are fewer lookups. Also, if the application uses [Bootsnap](https://github.com/Shopify/bootsnap), that saves the library from building unnecessary indexes, leading to lower memory usage.
```

=> 70136284426020
```

> As can be seen, the class object stored in the `User` constant is different after reloading.

Reviewer: ```suggestion
As indicated by the change in `object_id`, the class object stored in the `User` constant is different after reloading.
```

a reloaded `ApiGateway` would have a `nil` endpoint, because the code above does not run again.

> It is still possible set things up during boot, but they need to be wrapped in a `to_prepare` block, which runs on boot, and after each reload:

Reviewer: ```suggestion
It is still possible to set things up during boot, but any setup will need to be wrapped in a `to_prepare` block, which runs on boot, and after each reload:
```

By default, Rails uses `String#camelize` to know which constant a given file or directory name should define. For example, `posts_controller.rb` should define `PostsController` because that is what `"posts_controller".camelize` returns.

> It could be the case that some particular file or directory name does not get inflected as wanted. For instance, `html_parser.rb` is expected to define `HtmlParser` by default. What if `HTMLParser` is preferred? There are a few ways to customize this.

Reviewer: ```suggestion
It could be the case that some particular file or directory name does not get inflected as desired. For instance, `html_parser.rb` is expected to define `HtmlParser` by default. What if `HTMLParser` is preferred? There are a few ways to customize this.
```

The best way to follow what the loaders are doing is to inspect their activity.

The easiest way to do that is to include:

```ruby
Rails.autoloaders.log!
```

in `config/application.rb` after loading the framework defaults. That will print traces to standard output.

If logging to a file is preferred, configure this instead:

```ruby
Rails.autoloaders.logger = Logger.new("#{Rails.root}/log/autoloading.log")
```

> The Rails logger is not yet available when `config/application.rb` executes. If Rails logger is preferred, configure this setting in an initializer instead:

Reviewer: ```suggestion
The Rails logger is not yet available when `config/application.rb` executes. If the Rails logger is preferred, configure this setting in an initializer instead:
```
